compendium/DeclarativeServices/src/manager/ComponentFactoryImpl.cpp

```diff
@@ -0,0 +1,143 @@
+
+        ComponentFactoryImpl::ComponentFactoryImpl(
+            cppmicroservices::BundleContext const& context,
+            std::shared_ptr<cppmicroservices::logservice::LogService> logger,
+            std::shared_ptr<cppmicroservices::async::AsyncWorkService> asyncWorkSvc,
+            std::shared_ptr<SCRExtensionRegistry> extensionReg)
+            : bundleContext(context)
+            , logger(std::move(logger))
+            , asyncWorkService(asyncWorkSvc)
+            , extensionRegistry(extensionReg)
+        {
+            if (!bundleContext || !(this->logger) || (!this->asyncWorkService) || (!this->extensionRegistry))
+            {
+                throw std::invalid_argument("ComponentFactoryImpl Constructor "
+                                            "provided with invalid arguments");
+            }
+        }
+
+
+        void
+        ComponentFactoryImpl::CreateFactoryComponent(std::string const& pid,
+                                                     std::shared_ptr<ComponentConfigurationImpl>& mgr,
+                                                     cppmicroservices::AnyMap const& properties)
+        {
+            // Create the virtual metadata for the factory instance. 
+            // Start with the metadata from the factory 
+           auto newMetadata = std::make_shared<ComponentMetadata>(*mgr->GetMetadata());
+
+            newMetadata->name = pid;
+            // this is a factory instance not a factory component
+            newMetadata->factoryComponentID = "";
+
+            // Factory instance is dependent on the factory instance pid
+            newMetadata->configurationPids.clear();
+            newMetadata->configurationPids.emplace_back(pid);
+
+            // Look for dynamic targets in the references.
+            // A dynamic target will appear in the properties for the configuration object
+            // with the referenceName.target as the key and the target as the value. 
+            for (auto& ref : newMetadata->refsMetadata)
+            {
+                auto target = ref.name + ".target";
+                auto iter = properties.find(target);
+                if (iter != properties.end()) {
+                    // This reference has a dynamic target
+                    ref.target = cppmicroservices::ref_any_cast<std::string>(iter->second);
+                    // Verify that the ref.target is a valid LDAPFilter
+                    try
+                    {
+                        LDAPFilter filter(ref.target);
+                    }
+                    catch (std::exception const& e)
+                    {
+                        logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_ERROR,
+                                    "CreateFactoryComponent failed because of invalid target ldap filter"
+                                        + newMetadata->name + " target= " + ref.target);
+                        throw std::invalid_argument(e.what());
+                    }
```

Comment:
why not let the LDAPFilter constructor throw to the caller instead of catching and throwing (which is a performance hit)? The caller can log the exception.

Reply:
If the LDAPFilter is invalid, a std::exception will be thrown and the caller will log whatever message is included in the std::exception. I thought that it would be better to catch and log this here because we can include a more helpful message in the log. I think this is an error that will happen a lot to our users(it happened to me a few times) and warrants a bit more help.

Reply:
what is the message thrown from LDAPFilter?